ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,181 @@
+
+  // Constructor
+  public S3BatchDeleteHandler(DeleteBlobHandler deleteBlobHandler, FrontendMetrics frontendMetrics) {
+    this.deleteBlobHandler = deleteBlobHandler;
+    this.metrics = frontendMetrics;
+    this.xmlMapper = new XmlMapper();
+  }
+
+  /**
+   * Handles the S3 request and constructs the response.
+   *
+   * @param restRequest         the {@link RestRequest} that contains the request headers and body.
+   * @param restResponseChannel the {@link RestResponseChannel} that contains the response headers and body.
+   * @param callback            the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException exception when the processing fails
+   */
+  @Override
+  protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> callback) throws RestServiceException {
+
+    // Create the channel to read the request body
+    RetainingAsyncWritableChannel channel = new RetainingAsyncWritableChannel();
+
+    // inject metrics for batch delete
+    RestRequestMetrics requestMetrics =
+        metrics.deleteBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), false);
+    restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+
+    // Pass the callback to handle the response
+    restRequest.readInto(channel,
+        parseRequestBodyAndDeleteCallback(channel, restRequest, deleteBlobHandler, restResponseChannel, callback));
+  }
+
+  /**
+   * This method will read the request body, deserialize it, and trigger the batch delete process.
+   */
+  private Callback<Long> parseRequestBodyAndDeleteCallback(RetainingAsyncWritableChannel channel,
+      RestRequest restRequest, DeleteBlobHandler deleteBlobHandler, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> finalCallback) {
+
+    return buildCallback(metrics.s3BatchDeleteHandleMetrics, bytesRead -> {
+      if (bytesRead == 0) {
+        LOGGER.error("failed to read request into channel");
+        throw new RestServiceException("bytesRead is empty", RestServiceErrorCode.BadRequest);
+      }
+
+      S3MessagePayload.S3BatchDeleteResponse resp = new S3MessagePayload.S3BatchDeleteResponse();
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      ReadableStreamChannel readableStreamChannel =
+          new ByteBufferReadableStreamChannel(ByteBuffer.wrap(outputStream.toByteArray()));
+
+      // Deserialize the request body into a S3BatchDeleteObjects
+      deleteRequest = deserializeRequest(channel);
+
+      // Extract request path
+      RequestPath requestPath = (RequestPath) restRequest.getArgs().get(InternalKeys.REQUEST_PATH);
+      List<CompletableFuture<Void>> deleteFutures = new ArrayList<>();
+
+      if (deleteRequest.getObjects().size() > maxBatchSize) {
+        LOGGER.error("Exceeded batch size");
```

Comment:
A more informative errmsg could be 
```
LOGGER.error("Exceeded maximum batch size of ", MAX_BATCH_DELETE_SIZE);
```
Create one string and use it in both places - to print log and to pass in the RestServiceException